packages/iov-ethereum/src/ethereumcodec.ts

```diff
@@ -170,43 +173,71 @@ export class EthereumCodec implements TxCodec {
         options => options.contractAddress.toLowerCase() === toChecksummedAddress(json.to).toLowerCase(),
       );
 
-    let transaction: SendTransaction | SwapOfferTransaction;
+    let transaction: SendTransaction | SwapOfferTransaction | SwapClaimTransaction | SwapAbortTransaction;
     if (atomicSwap) {
-      const positionOpenMethodEnd = 4;
-      const positionOpenSwapIdBegin = positionOpenMethodEnd;
-      const positionOpenSwapIdEnd = positionOpenSwapIdBegin + 32;
-      const positionOpenRecipientBegin = positionOpenSwapIdEnd;
-      const positionOpenRecipientEnd = positionOpenRecipientBegin + 32;
-      const positionOpenHashBegin = positionOpenRecipientEnd;
-      const positionOpenHashEnd = positionOpenHashBegin + 32;
-      const positionOpenTimeoutBegin = positionOpenHashEnd;
-      const positionOpenTimeoutEnd = positionOpenTimeoutBegin + 32;
+      const positionMethodIdBegin = 0;
+      const positionMethodIdEnd = positionMethodIdBegin + 4;
+      const positionSwapIdBegin = positionMethodIdEnd;
+      const positionSwapIdEnd = positionSwapIdBegin + 32;
 
-      const swapId = input.slice(positionOpenSwapIdBegin, positionOpenSwapIdEnd) as SwapIdBytes;
-      const recipientChecksummedAddress = toChecksummedAddress(
-        Abi.decodeAddress(input.slice(positionOpenRecipientBegin, positionOpenRecipientEnd)),
-      );
-      const hash = input.slice(positionOpenHashBegin, positionOpenHashEnd) as Hash;
-      const timeoutHeight = new BN(input.slice(positionOpenTimeoutBegin, positionOpenTimeoutEnd)).toNumber();
+      const method = Abi.decodeMethodId(input.slice(positionMethodIdBegin, positionMethodIdEnd));
+      const swapId = input.slice(positionSwapIdBegin, positionSwapIdEnd) as SwapIdBytes;
 
-      transaction = {
-        kind: "bcp/swap_offer",
-        creator: creator,
-        swapId: swapId,
-        fee: fee,
-        amounts: [
-          {
-            quantity: decodeHexQuantityString(json.value),
-            fractionalDigits: constants.primaryTokenFractionalDigits,
-            tokenTicker: constants.primaryTokenTicker,
+      if (method === SwapContractMethod.Open) {
```

Comment:
I think a switch over `method` is slightly easier to follow, but this is good as well

Reply:
Yeah I thought about that too. I'll add it in the next PR.